### DISCUSSION TITLE
Is it possible to derive a binding using a key-path?

### DISCUSSION THREAD

**Discussion by wltrup** [pointfreeco/swift-composable-architecture#3196] [Q&A] [answered]
Hello 👋 

Suppose I have a feature state such as
```
struct State {
   var property1: String
   var property2: String
   var property3: String
}
```

In the feature's view I can derive a binding to each of the properties by using
```swift
$store.property1.sending(\.someAction)
```

with analogous lines for properties 2 and 3 but can I do this more generically, by referencing those properties and actions through key-paths, something along the lines of one of these?
```swift
$store[keyPath: propertyKeyPath].sending(actionCasePath)
$store.state[keyPath: propertyKeyPath].sending(actionCasePath)
```

I've managed to abstract away the action but not the state. It seems that I can only derive a binding by explicitly specifying the property I want a binding for.

Thanks in advance for any replies.

**Comment by mbrandonw** [comment, marked as answer]
Hi @wltrup, the syntax `$store.property1` is unlocked by `@dynamicMemberLook`, and technically desugars to:

```swift
$store[dynamicMember: \.property1]
```

So you should be able to do:

```swift
$store[dynamicMember: propertyKeyPath].sending(actionCasePath)
```

Can you let me know if that works?

**Reply by wltrup** [reply]
Hi @mbrandonw, thanks for the quick reply. Yes, that works.

**Reply by larryonoff** [reply]
@mbrandonw I once had a similar question, too. Looks that it would be great adding the answer to the docs, in a section about bindings.

**Reply by mbrandonw** [reply]
I'm just not sure where those docs could be put that people would find when they have this question. And also this isn't really a question about the library, but rather about Swift itself. If you find a place you think would be helpful we will think about it.

**Reply by larryonoff** [reply]
@mbrandonw looks that the best place is https://github.com/pointfreeco/swift-composable-architecture/blob/main/Sources/ComposableArchitecture/Documentation.docc/Articles/Bindings.md?plain=1#L97

adding something like
```swift
    Toggle(
      "Beer Bottle Opened",
      isOn: $store[dynamicMember: \.beerBottle.isOpen].sending(\.setBeerBottleOpen)
    )
```

BTW. I thought about a nested state property

**Reply by mbrandonw** [reply]
Well, that is a bit of a contrived example. There's no reason to do that over `$store.beerBottle.isOpen`. The only time to use `[dynamicMember:]` is when writing library code that needs to abstractly work with store bindings.

**Reply by larryonoff** [reply]
@mbrandonw Indeed! I got your idea. But do you mean swift will compile the line `$store.beerBottle.isOpen.sending(\.setBeerBottleOpen)`?

**Reply by mbrandonw** [reply]
Yes, that should compile just fine. Is it not for you?

**Reply by larryonoff** [reply]
I thought that it was not. But I don't have example yet. I'll create the issue if it will arise again

**Reply by mbrandonw** [reply]
It should work just fine since the thing you get back from `$store.beerBottle` has dynamic member lookup itself, allowing you to further chain into sub-properties:

https://github.com/pointfreeco/swift-composable-architecture/blob/15d2b12c84dd0b1c9c611df3e4be7b9641fb40f2/Sources/ComposableArchitecture/Observation/Binding%2BObservation.swift#L137-L149

**Reply by wltrup** [reply]
> The only time to use [dynamicMember:] is when writing library code that needs to abstractly work with store bindings.

That was exactly the reason I asked the question and the use-case I needed to implement in my current project.
